### PR TITLE
Working proof of concept for ES8388

### PR DESCRIPTION
**Pull request by troyhacks**
Lobbing this back to you for experimentation. Pins are hard-coded at the moment in places - like I2S init things. 

There's also lots of commented-out code that shows work towards eliminating the library. I didn't get it to work as well as the library (sometimes not at all), so more experimentation needed.

The order of how the ES8388 is initialized seems to matter greatly.

### CODE REVIEW

**Comment by troyhacks** [comment]
Should be in a "finished" state now. 

Still work to do on the actual ES8388 config and figure out all its registers.

File: usermods/audioreactive/audio_source.h [hunk 1]
```diff
@@ -490,75 +490,139 @@ class ES7243 : public I2SSource {
     int8_t pin_ES7243_SCL;
 };
 
+/* ES8388 Microphone
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h]: Not just microphone

File: usermods/audioreactive/audio_source.h [hunk 2]
```diff
@@ -490,75 +490,139 @@ class ES7243 : public I2SSource {
...
+   This is an I2S microphone that requires ininitialization over
+   I2C before I2S data can be received
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h]: Is this also accurate statement for his this chip or copy and paste?

File: usermods/audioreactive/audio_source.h [hunk 3]
```diff
@@ -490,75 +490,139 @@ class ES7243 : public I2SSource {
...
-        reg = es8388.readAllReg();
-        DEBUGSR_PRINTF("Reg-%02d = 0x%02x\r\n", i, reg[i]);
+    }
+
+    void _es8388I2cWrite(uint8_t reg, uint8_t val) {
+#ifndef ES8388_ADDR
+      Wire.beginTransmission(0x10);
+      #define ES8388_ADDR 0x10   // default address
+#else
+      Wire.beginTransmission(ES7243_ADDR);
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h]: Copy and paste error, incorrect constant for default address

File: usermods/audioreactive/audio_source.h [hunk 4]
```diff
@@ -490,75 +490,139 @@ class ES7243 : public I2SSource {
...
     }
 
-public:
-    ES8388Source(int sampleRate, int blockSize, int sdaPin, int sclPin) :
-      I2SSource(sampleRate, blockSize) {
+    void _es8388InitAdc() {
+      // This is by no means 100% figured but it's working for line-in
+      // with a little too much noise for my liking...
+      _es8388I2cBegin();
+      _es8388I2cWrite(0x08,0x00);
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h]: This section could really benefit for some comments, maybe not every value, but just which group of values are which, e.g input selection, i2s etc

File: usermods/audioreactive/audio_source.h [hunk 5]
```diff
@@ -490,75 +490,139 @@ class ES7243 : public I2SSource {
...
+        ERRORSR_PRINTF("\nAR: invalid I2S pin: SCK=%d, MCLK=%d\n", i2sckPin, mclkPin); 
+        return;
+      }
+
       // Reserve SDA and SCL pins of the I2C interface
-      PinManagerPinType pins[2] = { { pin_ES8388_SDA, true }, { pin_ES8388_SCL, true } };
-      if (!pinManager.allocateMultiplePins(pins, 2, PinOwner::HW_I2C)) {
-        pinManager.deallocateMultiplePins(pins, 2, PinOwner::HW_I2C);
-        ERRORSR_PRINTF("\nAR: Failed to allocate ES8388 I2C pins: SDA=%d, SCL=%d\n", pin_ES8388_SDA, pin_ES8388_SCL); 
+      PinManagerPinType es8388Pins[2] = { { sdaPin, true }, { sclPin, true } };
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h:580]: Should we be registering the I2S pins as well?

File: usermods/audioreactive/audio_source.h [hunk 6]
```diff
@@ -15,7 +15,7 @@
 #else
 #define SRate_t int
 #endif
-#include "ES8388.h"
+// #include "ES8388.h"
```
Comment (netmindz) [inline usermods/audioreactive/audio_source.h]: Remove entirely rather than comment and also remove from platformio.ini

File: platformio.ini [hunk 7]
```diff
@@ -587,7 +587,6 @@ build_flags_min =
 
 lib_deps_min =
   https://github.com/kosme/arduinoFFT#develop @ 1.9.2 ; used for USERMOD_AUDIOREACTIVE
-  https://github.com/netmindz/es8388arduino.git @ 0.0.1 ; used for USERMOD_AUDIOREACTIVE ES8388 source
```
Comment (netmindz) [inline platformio.ini]: Not sure why this isn't showing in the final PR. 

Have we got a duplicate entry in the file?

File: usermods/audioreactive/audio_source.h (hunk 5 above)
Comment (troyhacks) [reply usermods/audioreactive/audio_source.h:580]: They are registered when you init the I2S device.

File: platformio.ini (hunk 7 above)
Comment (troyhacks) [reply platformio.ini]: I think it was just out of sync

File: usermods/audioreactive/audio_source.h (hunk 4 above)
Comment (troyhacks) [reply usermods/audioreactive/audio_source.h]: Added and optimized. Still working on this in general to see if I can't get the line noise down.

File: usermods/audioreactive/audio_source.h (hunk 6 above)
Comment (troyhacks) [reply usermods/audioreactive/audio_source.h]: Fixed.

**Pull request merged**